Feature/vsc rproject (#482)
hpcleuven/VscDocumentation: source/compute/software/r_package_management.rst

you will need to install those packages locally. Do not hesitate to contact
your local support team when encountering issues along the way.

The sections below explain different ways to go about these installations.
These assume that you are either using a centrally installed R module
(:ref:`Using RStudio Projects<r_package_management_with_vsc_rproject>`,
:ref:`Standard R package installation<r_package_management_standard_lib>`)
or Conda (:ref:`Installing R packages using Conda<r_package_management_conda>`).
In all cases, however, it best to keep the following in mind:

* By default, R uses the `$VSC_HOME/R` path to install new packages.
  Since `$VSC_HOME` has limited quota, it is not the recommended location
  to install software. Instead, we recommend to use `$VSC_DATA`.

* R packages often include extensions written in compiled languages (e.g.
  C++ or Fortran). The centrally installed R modules are configured to compile
  these extensions with optimizations for the CPU architecture at hand
  (the ``"-march=native"`` option when using GCC compilers).
  This means that R extensions compiled in this way may not work

>   on other partitions than the one they were created on.

Author: I suggest we keep it as it is for now.

>   on other partitions than the one they were created on.

Reviewer: OK

* A given version of an R package may only work with certain R versions.
  For example, a package installed with R 4.2.2 may not work when used with
  R 4.4.1. Since typically many different R versions are available,
  as a user you should always be aware of the correct version to use for
  your project(s).

.. _r_package_management_with_vsc_rproject:

Using RStudio Projects with vsc-Rproject
----------------------------------------

When using a centrally installed R module, you can easily manage your own
R environments with custom packages using so-called
`RStudio Projects <https://docs.posit.co/ide/user/ide/guide/code/projects.html>`_.
RStudio Projects provide a self-contained, organized environment in R. Each project has
its own working directory, workspace and history which helps to avoid conflicts between different
projects. This structure encourages best practices such as using relative paths
and version control (e.g. git).

The :ref:`vsc-Rproject tool <vsc-Rproject>` helps you create RStudio Projects
and allows to compile extensions in a more portable way. The tool furthermore
simplifies setting up your R environment (including selecting the correct R

> module). Once activated, commands such as `install.packages(...)` or

Reviewer: ```suggestion
module). Once activated, commands such as ``install.packages(...)`` or
```

> `devtools::install_github(...)` will yield portable installations in a

Reviewer: ```suggestion
``devtools::install_github(...)`` will yield portable installations in a
```

well-defined, project specific location.

.. _r_package_management_standard_lib:

Standard R package installation
-------------------------------

> In case `vsc-Rproject` does not suit your needs, then you can still go about

Reviewer: This sentence does not help *promoting* the vsc-Rproject at all, I reckon. Why shall a user come to the conclusion that the standard R installation suits him better than using vsc-Rproject?

You need to start this subsection like the following:

"In order to install your desired package(s) on top of an R module, take the following steps."
... or something similar.

Author: How does this sound?

```suggestion
If, instead of relying on RStudio Projects to manage isolated libraries for each project, you’d rather maintain a single, centralized R package library, then you can still go about
```

Reviewer: Let me reword your own sentence:

"Instead of relying on RStudio Projects to manage isolated libraries for each project, it is a better practice to maintain a single, centralized R package library."

Author: We could also just ignore the RStudio aspect here and perhaps say something along the lines of:

```suggestion
If your goal is to create a single, centralized R package library, then you can go about
```

Reviewer: Sounds good to me.

your local package installations as follows.

> Considering the possible dependencies on the OS version, CPU microarchitecture

Reviewer: ```suggestion
Considering the dependence of the compiled package on the host OS version, CPU microarchitecture
```

> and R version (see the :ref:`Intro<r_package_management_intro>`), we recommend

Reviewer: ```suggestion
and R version (see :ref:`above<r_package_management_intro>`), we recommend
```

.. code:: r

       $ module load R-bundle-Bioconductor/3.16-foss-2022b-R-4.2.2

It is possible, however, that these modules do not contain all R packages you need
or that the package versions do not meet your requirements. In this case you will
need to locally install those packages, as will be described below.

Do not hesitate to contact your local support team when encountering issues during these local installations.

.. _r_package_management_with_vsc_rproject:

RStudio Projects and HPC
------------------------

If you need a custom R environment (`e.g.` with R packages not provided by modules),
you can easily manage such environments with so-called
`RStudio Projects <https://docs.posit.co/ide/user/ide/guide/code/projects.html>`_.
RStudio Projects provide a self-contained, organized environment in R. Each project has
its own working directory, workspace and history which helps to avoid conflicts between different
projects. This structure encourages best practices such as using relative paths
and version control (e.g. git).

However, using RStudio Projects on a heterogenous HPC system posses some challenges.

> The first challenge arrises when installing packages in your project's package library.

Author: The latest changes to the r_package_management.rst page made the above suggestion outdated.

By default, these R packages are compiled with ``"-march=native"``.
As a result, the package will be installed in a way that is optimal for the CPU microarchitecture
of the node used for the installation. However, when you try to use that same package on
a node with a different architecture, it may give worse performance or even break entirely.

It is also important to be aware that R packages are version specific. A package installed for
``R/4.2.2`` may not work when used with ``R/4.4.1`` or visversa. Since there are many R versions
available on our HPC system, as a user you should always be aware of which version to use for which project.

> With these difficulties in mind, we have developed `vsc-Rproject` which provides a convenient

Author: The latest changes to the r_package_management.rst page made the above suggestion outdated.

    Usage of conda is discouraged in the clusters at UAntwerpen, UGent,
    and VUB.

> The paragraphs below illustrate how to install and use R and R packages

Reviewer: Two more requests:
- please check that everywhere, Conda is mentioned with capital-C (official name)
- starting from line 187, I propose the following change for better formatting:

If the result is blank, or reports that Conda can not be found, modify the ``PATH`` environment variable appropriately by adding miniconda’s ``bin`` directory to ``PATH``.

   $ conda search -c conda-forge r-base  # select one of available versions for the step below
   $ conda create -n science -c conda-forge r-base=<version> r-essentials

This command creates a new conda environment called "science", and installs your prefered R
version from the conda-forge channel as well as the r-essentials bundle which includes number

> of commonly used R packages such as ggplot2, glmnet, dplyr, tidyr, and shiny.

Reviewer: ```suggestion
of commonly used R packages such as ``ggplot2``, ``glmnet``, ``dplyr``, ``tidyr``, and ``shiny``.
```
